pkg/store/cache/caching_bucket_factory.go

```diff
@@ -104,10 +104,17 @@ func NewCachingBucketFromYaml(yamlContent []byte, bucket objstore.Bucket, logger
 	case string(GroupcacheBucketCacheProvider):
 		const basePath = "/_galaxycache/"
 
-		c, err = cache.NewGroupcache(logger, reg, backendConfig, basePath, r, bucket,
-			isTSDBChunkFile, isMetaFile, isBlocksRootDir,
-			config.MetafileExistsTTL, config.MetafileDoesntExistTTL, config.MetafileContentTTL,
-			config.ChunkObjectAttrsTTL, config.ChunkSubrangeTTL, config.BlocksIterTTL)
+		groupcacheCfg := cache.NewCachingBucketConfig()
+
+		// Configure cache.
+		groupcacheCfg.CacheGetRange("chunks", c, isTSDBChunkFile, config.ChunkSubrangeSize, config.ChunkObjectAttrsTTL, config.ChunkSubrangeTTL, config.MaxChunksGetRangeRequests)
```

Comment:
Let's create `groupcacheCfg` / `cfg` once? (:

Reply:
I think we need two `cfg`'s. One for groupcache that will accept a `nil` `cache.Cache` and then one that will accept a cache provided by `NewTracingCache`.

Reply:
Mhm, that's true, but I'd suggest ruthlessly refactoring this to avoid confusion in the future. We care about long-term maintainability. How about creating a new method such as:

```go
func (cfg *CachingBucketConfig) SetCacheImplementation(c Cache)
```

That would iterate over all configurations and set `OperationConfig.Cache` on each of them? Then, we could construct one `cache.NewCachingBucketConfig` with `nil` as the `cache.Cache` parameter, and finally call `cfg.SetCacheImplementation(c)` after `c = cache.NewTracingCache(c)`. This should effectively decouple `CachingBucketConfig` from the group cache. Downstream users such as Cortex wouldn't need to change anything as well :thinking:

Reply:
I don't understand still. With this we just won't have to create 2 `cfg`'s, so why won't Cortex change anything? We are still refactoring the TTLs into a struct. 😕

Reply:
We need to use the struct `CachingBucketConfig` [here].(https://github.com/GiedriusS/thanos/blob/088b9d338e9dc2975ed3b11be2b2e4be0d0ef200/pkg/cache/groupcache.go#L82)

Reply:
Yes, my bad, maybe I have expressed myself improperly. Don't mind the last sentence there :joy: my suggestion is about making the code less error-prone for ourselves so that we don't have to configure it twice :smile:

Reply:
Okay, got it.